### DISCUSSION TITLE
Mesa Development Session Notes (December 2021)

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#1093] [General]
## Tasks to Complete by Next Dev Session (January 2022): 

Interested in helping develop Mesa, come to the [next dev session](https://github.com/projectmesa/mesa/discussions/1094)

**Pull Request Review: ** (request #, person, difficulty, category)
[1091 tpike3](https://github.com/projectmesa/mesa/pull/1091) Easy Docs **COMPLETE**
[1090 tpike3](https://github.com/projectmesa/mesa/pull/1090) Easy Docs **COMPLETE**
[1087 jackiekazil](https://github.com/projectmesa/mesa/pull/1087) Significant Backward compatibility  **COMPLETE**
[1084 jackiekazil](https://github.com/projectmesa/mesa/pull/1084) Significant Backward compatibility **COMPLETE**
[1083 danielxu05](https://github.com/projectmesa/mesa/pull/1083) Medium Checks
[1075 tpike3](https://github.com/projectmesa/mesa/pull/1075) Easy Check Fail **COMPLETE**
[1057 tpike3](https://github.com/projectmesa/mesa/pull/1057) Easy Example merge 
[1035 rht](https://github.com/projectmesa/mesa/pull/1035) TBD Checks
[998 majdal](https://github.com/projectmesa/mesa/pull/998) Easy Docs
[991 jackiekazil](https://github.com/projectmesa/mesa/pull/991)TBD validate
[924 rht and tpike3](https://github.com/projectmesa/mesa/pull/924) significant validate/backward compatibility **MERGED--Moved to 1108 for docs update and deprecation warning**

### Scalability: 
Cython implementation (rht, tpike3, danielxu5)
- Best Approach?: build a separate Cython based Mesa  or build a conversion module
- GPUs: Can Cython approach also allow for GPU usage   

### Education
- Talk with Santa Fe Institute about building a Complexity Explorer based tutorial (tpike3)

**Comment by rht** [comment]
spaCy has a note on how it uses Cython: https://spacy.io/api/cython. The quickest win we can get is maybe to cythonize the `Grid` class:
- Faster than using Numpy `ndarray` on small array size. See https://stackoverflow.com/questions/6559463/why-is-numpy-array-so-slow
- The GIL can be disabled

One drawback with using Cython is that it won't work in PyPy.

**Comment by jackiekazil** [comment]
@tpike3 Ty for posting these!

**Comment by jackiekazil** [comment]
#991 needs a conflict resolution. I can either do that OR someone else can and I can evaluate. I don't think I can should be doing it and merging. Not sure if @Corvince has bandwidth these days -- @Corvince, please comment on the ticket if you do.

**Reply by Corvince** [reply]
Hey there, thanks for pinging me, I wouldn't have noticed otherwise. 
Actually it's a bit better these days for me so hopefully I can fix the conflicts the next days

**Reply by jackiekazil** [reply]
<3 Awesome! Let me know if you need any help!

**Comment by tpike3** [comment]
SFI [Complexity Explorer](https://www.complexityexplorer.org/home) is excited about a Mesa tutorial. I will fill out the attached questionnaire and share. Next step will be to have a Zoom call and start talking specifics.  

[ComplexityExplorer Online Course Proposal.txt](https://github.com/projectmesa/mesa/files/7664229/ComplexityExplorer.Online.Course.Proposal.txt)

**Comment by EwoutH** [comment]
There are currently at least 3 PRs open that would reduce backward compatibility. I would propose:

1. Merging as much as possible bugfixes and improvements that don't break backwards compatibility
2. Tagging and deploying a Mesa 0.8.10 release
3. Merging the things that do break backwards compatibility
4. Tagging a 0.9.0 release

[GitHub Milestones](https://github.com/projectmesa/mesa/milestones) are a great way to keep track which issues and PRs are scheduled for which release. Creating Milestones for 0.8.10 and 0.9.0 could be very useful.

**Reply by rht** [reply]
Would be great if @EwoutH's PRs that don't break compatibility can be reviewed and merged soon. Otherwise, there is this psychological barrier that any contribution one make take weeks to be merged.

Can be merged IMO:
- #1099
- #1097
- #1091
- #1090

**Reply by tpike3** [reply]
Concur and done!

**Comment by tpike3** [comment]
I filled out the complexity explorer questionnaire (attached). Please feel free to provide feedback, I will send out before the new year.  

[Complexity Explorer Proposal.pdf](https://github.com/projectmesa/mesa/files/7762890/Complexity.Explorer.Proposal.pdf)
(changed to pdf here too)

**Reply by rht** [reply]
The file is in Windows line endings. I tried to convert it to Unix using `dos2unix` to no avail. I think it is better if you format it as pdf.

**Reply by tpike3** [reply]
Arrgh...please see attached .pdf
[Complexity Explorer Proposal.pdf](https://github.com/projectmesa/mesa/files/7762890/Complexity.Explorer.Proposal.pdf)

**Reply by rht** [reply]
- Typo: should be "agent-based model" instead of "agent based model"
- For bullet point 12, I think an assessment for the part 2 of the course should be a merged pull request? There are plenty of NetLogo examples that can be ported to Mesa. See https://github.com/NetLogo/models/tree/master/Sample%20Models

**Reply by rht** [reply]
See also #134.

**Reply by jackiekazil** [reply]
LGTM. 
@rht is right on the separate repo. We need to prioritize next dev sprint.

**Comment by tpike3** [comment]
Interesting, so to makes sure I am understanding..... part 1 of the tutorial could be building an model (such as from nlogo examples) and then part 2 is running through a merge request to add to Mesa model library>

Then their assessment is building their own example model and doing a merge into a Mesa examples folder?

I like it.....that is a rock solid assessment. 

(I will also fix the typo)

**Reply by EwoutH** [reply]
Great idea for the PR @rht! Maybe even an (optional) part can be included to review or improve someone others model.

**Reply by rht** [reply]
Yeah for part 2, but for part 1 you need to start with introducing using the wolf-sheep model. There is an issue that the participant needs to be able to extract model implementation detail from the NetLogo syntax. This is confusing, especially that the model code and the GUI code are mixed into 1 file, e.g. https://github.com/NetLogo/models/blob/master/Sample%20Models/Philosophy/Signaling%20Game.nlogo. This is not a problem for a semester-long course, though. @tpike3 what is the duration of the course?

Another problem is that, then it becomes a course in *both* Mesa and NetLogo.

**Reply by rht** [reply]
I think we can alleviate the burden of the participants by asking them to only read the textual model description instead of the NetLogo code. Then we can review and be the one to do the comparison with the actual NetLogo code.

**Reply by tpike3** [reply]
@rht The idea is to be a tutorial of 4-6 hours. I figure only some people will do the assessments but typical rule of thumb is 2 hours of work per hour of class so I think these models would be reasonable.  Here is the [Complexity Explorer](https://www.complexityexplorer.org/courses) courses.

Per your point, my understanding is that the netlogo repo is just a comprehensive set of models so participants can mimic a model in python without the harder task of trying to come up with something valid to model. These would give them a list of time tested models.

**Reply by rht** [reply]
OK that makes sense.

**Reply by rht** [reply]
IMO #134 needs to be resolved before you can submit the proposal. My vote is to make a separate repo for the new example contributions.

**Reply by tpike3** [reply]
Roger, concur, I think a separate repo is the common consensus. It will have to be discussed as the next dev meeting. 

However, I think we can probably submit the proposal before the next dev meeting as the details don't have to be flushed out until we record (I am thinking the earliest is June/July 2022) . I do however need to update various questions for more robust assessments in which students build an independent example and then do a pull request to the example folder.

**Reply by tpike3** [reply]
Copy of submitted proposal to SFI. 

[Mesa Complexity Explorer Proposal.pdf](https://github.com/projectmesa/mesa/files/7799043/Mesa.Complexity.Explorer.Proposal.pdf)

**Reply by tpike3** [reply]
SFI has accepted and is starting to work their support piece.

**Comment by jackiekazil** [comment]
Help me out -- where are we with merges and what do we need to for the next release? (I lost track with the holiday.)

**Reply by tpike3** [reply]
The minimum I am tracking for the next release is the update to docs for [BatchRunner](https://github.com/projectmesa/mesa/issues/1109) which @jonahaag is working on.

The other one we may want to consider is the contribution by @rht that improves performance by [optimizing empties](https://github.com/projectmesa/mesa/pull/1116). 

My understanding is once we have a consensus self.empties it should be pretty quick

**Reply by jackiekazil** [reply]
So I should hold off until those 2 things come thru?

**Reply by tpike3** [reply]
I would say yes!
